### PR TITLE
[sdk] Add verifying and non-verifying client equivalence testing

### PR DESCRIPTION
**Pull request by phlip9**
### Overview

This PR adds a test framework for ensuring that verifying and non-verifying clients handle single requests and request batches identically (when pinned to the same version).

### [sdk] Add client support for sending MethodRequest enum

In the same way that we have `batch`, which takes a list of `MethodRequest` enums:

```rust
impl Client {
    pub fn batch(&self, requests: Vec<MethodRequest>)
        -> Result<Response<Vec<MethodResponse>>>;
}
```

this diff adds an additional method which takes only a single `MethodRequest` enum and returns a `MethodResponse` enum:

```rust
impl Client {
    pub fn request(&self, request: MethodRequest)
        -> Result<Response<MethodResponse>>;
}
```

While this method is less useful under normal circumstances (users will probably prefer to just call something like `get_account`, which correctly parses out the desired `AccountView` type), this method is instead more useful for programmatic or generic logic. For instance, fuzzing the client with random requests.

### [smoke-test] Add initial framework for verifying client equiv test

The idea here is to whitebox fuzz the VerifyingClient and non-verifying Client with the same random requests and assert that they both handle the same requests identically.

Ideally, we should be able to just add new cases to the `arb_request` strategy to test new request types as we add support for them.

### [smoke-test] Add verifying client batch request equiv testing

In addition to fuzzing single request -> response against the verifying and non-verifying client, this diff adds full batch request -> batch response fuzzing to ensure that both clients handle batches identically.

The only minor wrinkle is that we need to filter out random batches that would exceed the maximum server batch size after going through the VerifyingClient batch transform. Hence, this diff also adds a function `VerifyingClient::actual_batch_size` which computes the final batch size of a request batch, which we use to rejection sample batches under the limit.

### CODE REVIEW

**Comment by phlip9** [comment]
dropped the EventKey formatting commit, since it caused some problems with the language e2e tests

File: testsuite/smoke-test/src/verifying_client.rs [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    let arb_account = select(accounts.to_owned());
+    let arb_version = prop_oneof! [
+        5 => Just(current_state_version),
+        // query a historical version
+        20 => 0u64..current_state_version,
+        // occasionally choose an invalid version
+        1 => Just(u64::MAX / 2),
+    ];
+    (arb_account, arb_version)
+        .prop_map(|(address, version)| MethodRequest::get_account_by_version(address, version))
```
Comment (bmwill) [inline testsuite/smoke-test/src/verifying_client.rs:176]: So as we add more methods to the verifying client we can add more things here.

**Review by bmwill — COMMENTED** [review]
Really only 1 nit

File: sdk/client/src/blocking.rs [hunk 2]
```diff
@@ -121,6 +121,12 @@ impl BlockingClient {
         self.send_batch(requests)
     }
 
+    pub fn request(&self, request: MethodRequest) -> Result<Response<MethodResponse>> {
+        let method = request.method();
+        let resp: Response<serde_json::Value> = self.send(request)?;
+        resp.and_then(|json| MethodResponse::from_json(method, json).map_err(Error::decode))
+    }
```
[the remark this answers is not in the document]
Comment (bmwill) [reply sdk/client/src/blocking.rs:128]: I'm incorrect, this already does that validation.

**Review by bmwill — APPROVED** [review]
Took another look and realized I was mistaken. This looks great!

File: testsuite/smoke-test/src/verifying_client.rs (hunk 1 above)
Comment (phlip9) [reply testsuite/smoke-test/src/verifying_client.rs:176]: Yeah that's the idea, generate different request types here as we add support for them

**Pull request merged**